rllab/dynamics_randomization/mujoco_model_gen.py

```diff
@@ -0,0 +1,133 @@
+import atexit
+import queue
+from queue import Queue
+from threading import Event
+from threading import Thread
+
+from lxml import etree
+from mujoco_py import load_model_from_xml
+import numpy as np
+
+from rllab.dynamics_randomization import Distribution
+from rllab.dynamics_randomization import Method
+
+
+class MujocoModelGenerator:
+    """
+    A worker thread to produce to MuJoCo models with randomized dynamic
+    parameters, which are specified by the users of rllab with the class
+    Variations.
+    """
+
+    def __init__(self, file_path, variations):
+        """
+        Starts all the member fields of the class and the worker thread.
+        Parameters
+        ----------
+        file_path : string
+            The absolute path to the XML file that contains the MuJoCo
+            model.
+        variations: Variations
+            An list of Variation objects that indicate the dynamic parameters
+            to randomize in the XML file.
+        """
+        self._variations = variations
+        self._file_path = file_path
+        # Synchronized queue to store mujoco_models
+        self._models = Queue(maxsize=10)
+        # Worker Thread
+        self._worker_thread = Thread(
+            target=self._generator_routine, daemon=True, name="Worker-Thread")
+        # Reference to the generated model
+        self._mujoco_model = None
+        # Event to stop the worker thread
+        self._stop_event = Event()
+        atexit.register(self.stop)
+        self._worker_thread.start()
+
+    def get_model(self):
+        """
+        Gets the MuJoCo model produced by the worker thread in this class.
+        This call may block in case the calling thread asks for the model before
+        the worker thread has finished.
+        Returns
+        -------
+        PyMjModel
+            A MuJoCo model with randomized dynamic parameters specified by the
+            user in this class.
+        """
+        if not self._worker_thread.is_alive():
+            # If worker thread terminates because of an error, terminates main thread
+            raise ChildProcessError("Error raised in Worker-Thread")
+
+        try:
+            return self._models.get(timeout=1)
+        except queue.Empty:
```

Comment:
I would prefer that we raise a warning in the log and try again. You should only throw an uncaught exception if your program can literally not continue at all. But your program can just wait.

If the worker thread dies, that is catastrophic and will  probably raise some other exception in the worker thread. 

But consistently underflowing the queue is in many ways worse, because it makes things really slow, and we should be warning when this happens.